src/Hazelcast.Net/Serialization/Compact/CompactWriter.cs

```diff
@@ -0,0 +1,472 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#nullable enable
+
+using System;
+using System.Numerics;
+using Hazelcast.Core;
+using Hazelcast.Models;
+
+namespace Hazelcast.Serialization.Compact
+{
+    // notes:
+    // when passing a lambda as an Action<> parameter, the compiler generates a supporting
+    // class which statically caches the Action<> object so it's only instantiated once.
+
+    internal class CompactWriter : CompactReaderWriterBase, ICompactWriter
+    {
+        private readonly IWriteObjectsToObjectDataOutput _objectsWriter;
+        private readonly ObjectDataOutput _output;
+        private readonly int[]? _offsets;
+
+        private int _dataPosition;
+        private bool _completed;
+
+        public CompactWriter(IWriteObjectsToObjectDataOutput objectsWriter, ObjectDataOutput output, Schema schema)
+            : base(schema, output.Position)
+        {
+            _objectsWriter = objectsWriter ?? throw new ArgumentNullException(nameof(objectsWriter));
+            _output = output ?? throw new ArgumentNullException(nameof(output));
+
+            if (schema.HasReferenceFields) _offsets = new int[schema.ReferenceFieldCount];
```

Comment:
I see it's used with null forgiving operator all over there but `_offsets` is only initialized if `HasReferenceFields`. Isn't there is room for null exception  🤕?

Reply:
`_offsets` are only used if we have reference fields, so we should never use it if we don't have reference fields (i.e. when it's `null`). There is room for a nullref exception... if our code has bugs. We could test for `null` everywhere, but we'd just have to throw some sort of "internal error" exception... so I though, let's just keep the nullref.

Reply:
> _offsets are only used if we have reference fields

That's fair enough.